src/abstractdataframe/join.jl

```diff
@@ -81,16 +81,573 @@ end
+        end
+        if !already_joined
+            if right_shorter
+                if left_col isa AbstractVector{<:Union{Integer, Missing}} &&
+                   right_col isa AbstractVector{<:Union{Integer, Missing}}
+                    left_ixs, right_ixs = _innerjoin_unsorted_int(left_col, right_col)
+                else
+                    left_ixs, right_ixs = _innerjoin_unsorted(left_col, right_col)
+                end
+            else
+                if left_col isa AbstractVector{<:Union{Integer, Missing}} &&
+                   right_col isa AbstractVector{<:Union{Integer, Missing}}
+                    right_ixs, left_ixs = _innerjoin_unsorted_int(right_col, left_col)
+                else
+                    right_ixs, left_ixs = _innerjoin_unsorted(right_col, left_col)
+                end
+            end
+        end
+    end
+
+    dfl = joiner.dfl[left_ixs, :]
+    dfr_noon = joiner.dfr[right_ixs, Not(joiner.right_on)]
+
+    ncleft = ncol(dfl)
+    cols = Vector{AbstractVector}(undef, ncleft + ncol(dfr_noon))
+
+    for (i, col) in enumerate(eachcol(dfl))
+        cols[i] = col
+    end
+    for (i, col) in enumerate(eachcol(dfr_noon))
+        cols[i+ncleft] = col
+    end
+
+    new_names = vcat(_rename_cols(_names(joiner.dfl), left_rename, joiner.left_on),
+                     _rename_cols(_names(dfr_noon), right_rename))
+    res = DataFrame(cols, new_names, makeunique=makeunique, copycols=false)
+
+    return res, nothing, nothing
+end
+
+@inline function find_next_range(x::AbstractArray, start::Int, start_value)
+    stop_value = start_value
+    n = length(x)
+    stop = start + 1
+    while stop <= n
+        @inbounds stop_value = x[stop]
+        isequal(start_value, stop_value) || break
+        stop += 1
+    end
+    return stop, stop_value
+end
+
+function _innerjoin_sorted(left::AbstractArray, right::AbstractArray)
+    left_n = length(left)
+    right_n = length(right)
+
+    left_ixs = Int[]
+    right_ixs = Int[]
+
+    (left_n == 0 || right_n == 0) && return left_ixs, right_ixs
+
+    # lower bound assuming we get matches
+    sizehint!(left_ixs, min(left_n, right_n))
+    sizehint!(right_ixs, min(left_n, right_n))
+
+    left_cur = 1
+    left_val = left[left_cur]
+    left_new, left_tmp = find_next_range(left, left_cur, left_val)
+
+    right_cur = 1
+    right_val = right[right_cur]
+    right_new, right_tmp = find_next_range(right, right_cur, right_val)
+
+    while left_cur <= left_n && right_cur <= right_n
+        if isequal(left_val, right_val)
+            if left_new - left_cur == right_new - right_cur == 1
+                push!(left_ixs, left_cur)
+                push!(right_ixs, right_cur)
+            else
+                for (left_i, right_i) in Iterators.product(left_cur:left_new - 1,
+                                                           right_cur:right_new - 1)
+                    push!(left_ixs, left_i)
+                    push!(right_ixs, right_i)
+                end
```

Comment:
@nalimilan for this part of code I am also testing:
```
idx = length(left_ixs)
left_range = left_cur:left_new - 1
right_range = right_cur:right_new - 1
to_grow = Base.checked_mul(length(left_range), length(right_range))
Base._growend!(left_ixs, to_grow)
Base._growend!(right_ixs, to_grow)
@inbounds for (left_i, right_i) in Iterators.product(left_range, right_range)
    idx += 1
    left_ixs[idx] = left_i
    right_ixs[idx] = right_i
end
```
which should be faster (and in isolated tests it is faster), but it seems that in this code it is not faster unless we would create thousands of rows (i.e. `to_grow` would be very large).

This is strange. Do you see a reason for this?

Reply:
To be sure I understand correctly, the alternative code in your comment avoids resizing the vectors on each `push!` and instead resizes them once? Yes I'd expect this to be faster. Maybe try putting it in a separate function in case some type instability destroys the expected gain? Though it's not completely absurd to think that the gain only appears with a large number of rows (`push!` doubles the size of the array each time it needs to grow it, so it should be reasonably fast for a small number of rows.)

(BTW, I don't think you need `Base._growend!`, `resize!` should do the same thing.)

Reply:
You understand correctly.

`resize!` does an additional `if` check that cannot be optimized-out statically.

I will benchmark it again then.

Reply:
I have run the benchmarks again and in practical cases using `_growend!` it is ~10% faster. I have pushed the change to GitHub then.

I will merge this PR tomorrow if there will be no more comments on it.

Reply:
You mean 10% faster than the PR before that commit, or than using `resize!`? If possible I'd avoid relying on internal functions.

Reply:
In comparison to original commit. I will check `resize!`

Reply:
I have changed to `resize!`, as it has less 1 ms overhead over the original approach with `_growend!` for large tables.